File: lc-problems/410-Split-Array-Largest-Sum/sample_binary_search.cc
class Solution {
public:
    int splitArray( vector< int >& nums, int m ) {

        long long int ans = 0;
        int           n   = nums.size();
        for ( int i = 0; i < n; i++ )
            ans += nums[ i ];

        long long int l = 0;
        long long int r = ans;

        while ( l <= r ) {
            long long int mid = ( l + r ) / 2;
            // cout << l << " " << r << " " << mid << endl;
            int           cuts     = 0;
            long long int sum      = 0;
            bool          possible = true;
            for ( int i = 0; i < n; i++ ) {
                if ( nums[ i ] > mid ) {
                    possible = false;
                    break;
                }
                if ( sum + nums[ i ] <= mid )
                    sum += nums[ i ];
                else {
                    sum = nums[ i ];
                    cuts++;
                }
            }

            if ( cuts + 1 <= m && possible ) {
                r   = mid - 1;
                ans = min( ans, mid );
            }
            else {
                l = mid + 1;
            }
        }
        // cout << ans << endl;
        return ans;
    }
};

File: lc-problems/410-Split-Array-Largest-Sum/sample_binary_search.py
#!/usr/bin/env python


class Solution(object):
    def splitArray(self, nums, p):
        lo, hi = max(nums), sum(nums)

        while lo <= hi:
            mid = (lo + hi) / 2
            curr, partitions = 0, 1

            for num in nums:
                curr += num
                if curr > mid:
                    partitions += 1
                    curr = num

            if partitions <= p:
                hi = mid - 1
            else:
                lo = mid + 1

        return lo
